### PR TITLE
ActiveAE: transcode to E-AC3 when eac3passthrough is enabled

### PR DESCRIPTION
**Pull request by reardonia**
## Description
  Use the existing E-AC3 encoder support (added for webOS in #27541) for all
  platforms. When AC3 transcoding and E-AC3 passthrough are both enabled,
  automatically transcode to E-AC3 instead of AC3.

  ## Motivation and context
  The webOS E-AC3 transcoding added in #27541 is platform-specific, gated behind
  `WebOSTVPlatformConfig::SupportsEAC3()`. This change makes the same capability
  available on all platforms by using the existing `eac3passthrough` setting to
  determine codec selection — if the receiver supports E-AC3, use it.

  ~~Also raises E-AC3 encode bitrate from 768kbps to 1024kbps (tested maximum for
  the ffmpeg EAC3 encoder).~~

  Relates to #27991

  There are no new settings, which could be _slightly_ misleading as the existing transcode
  settings says "Transcode to AC3".  But I wanted to minimize the changes here.

  ## How has this been tested?
  Tested on Intel NUC (i3-8109U) with HDMI output to receiver supporting both
  AC3 and E-AC3 passthrough. Verified transcoding activates for multichannel
  content with both AC3 (eac3passthrough off) and E-AC3 (eac3passthrough on).
  Confirmed 1024kbps is the maximum working bitrate for the ffmpeg EAC3 encoder.

  ## What is the effect on users?
  Users with E-AC3-capable receivers will automatically get higher quality
  transcoded audio without any new settings. Behaviour is unchanged when
  E-AC3 passthrough is not enabled.

  ## Screenshots (if appropriate):

  ## Types of change
  - [ ] **Bug fix** (non-breaking change which fixes an issue)
  - [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
  - [X] **Improvement** (non-breaking change which improves existing functionality)
  - [ ] **New feature** (non-breaking change which adds functionality)
  - [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
  - [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
  - [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
  - [ ] **None of the above** (please explain below)

  ## Checklist:
  - [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project
  - [ ] My change requires a change to the documentation, either Doxygen or wiki
  - [ ] I have updated the documentation accordingly
  - [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
  - [ ] I have added tests to cover my change
  - [X] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp [hunk 1]
```diff
@@ -7,7 +7,7 @@
  */
 
 #define AC3_ENCODE_BITRATE 640000
-#define EAC3_ENCODE_BITRATE 768000
+#define EAC3_ENCODE_BITRATE 1024000
```
Comment (fritsch) [inline xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp]: Please leave this change out for now. We decided for the lower rate due to Netflix and Soundbar compatibility.

**Comment by fritsch** [comment]
The text for Dolby transcode should be adjusted so that user knows when he gets ac3 and when eac3.

If ffmpeg nowadays needs a forced flag or has one for eac3 encoder output enablement, please add this as well. Cause due to your change ffmpeg needs to be compiled with eac3 encoder output else we silently crash on transcode start.

**Comment by fritsch** [comment]
See, especially on windows, which has a dedicated --enable-encoder= list and eac3 is missing here.

Edit: Maybe to clarify. I am well aware that ffmpeg internally uses ac3parse, ac3 etc. for eac3 as well as ac3  - but as the option eac3 exists, we should properly supply it. Reason: they might split them up in "some future" - that way we have it properly right at the start.

**Comment by sundermann** [comment]
The encoder should probably check the return value of avcodec_find_encoder in the eac3 case and then fall back to ac3. That was not an issue on webOS because it's a depends only platform

File: xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp (hunk 1 above)
Comment (reardonia) [reply xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp]: Will do. I knew about Netflix, but is there a pointer to anything on that being a known limit on soundbars?  Curious because another media package I work with changed to this limit as maximally compatible but I'd like to update there as well if there is data showing otherwise.

**Comment by reardonia** [comment]
> The text for Dolby transcode should be adjusted so that user knows when he gets ac3 and when eac3.

I initially made ac3 vs eac3 user selectable, should I bring that back in or are you comfortable with auto-selecting?

File: xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp [hunk 2]
```diff
@@ -101,6 +101,15 @@ bool CAEEncoderFFmpeg::Initialize(AEAudioFormat& format, bool allow_planar_input
     m_CodecID = AV_CODEC_ID_EAC3;
     m_BitRate = EAC3_ENCODE_BITRATE;
     codec = avcodec_find_encoder(m_CodecID);
+    if (!codec)
```
Comment (sundermann) [inline xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp:98]: This needs to set`format.m_streamInfo.m_type` to AC3. Also this is duplicating the code below

**Comment by sundermann** [comment]
https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/21668 is actually bringing support for eac3 dependent streams so this change makes sense also in future.

https://github.com/sundermann/xbmc/commit/b011afb5d16b1072e6f4bc3a68392b286336fd89 has a quick and dirty patch to pull the PR in. It works perfectly on webOS to convert TrueHD to 7.1 eac3 (without ATMOS still)

**Comment by reardonia** [comment]
> https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/21668 is actually bringing support for eac3 dependent streams so this change makes sense also in future.
> 

Does the bitrate limit apply overall or per stream?  Obviously with dependent streams we'd want to go past 768kbps

**Comment by sundermann** [comment]
No idea, but I would still leave it as is in this PR

File: xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp (hunk 2 above)
Comment (reardonia) [reply xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp:98]: Yup, sorry, brain fry.  Made this cleaner by assuming that "eac3 passthrough" implies "ac3 passthrough", no need to look it up again, so we get rid of Settings here entirely.

Comment (sundermann) [reply xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp:98]: Not sure if we can safely drop the setting checks here. But I do agree that it would be better checking in ActiveAE.

Also it appears that there is no checks if EAC3 is actually enabled in ActiveAE.

**Review by sundermann — APPROVED** [review]
Looks good after the changes

**Comment by reardonia** [comment]
> See, especially on windows, which has a dedicated --enable-encoder= list and eac3 is missing here.
> 
> Edit: Maybe to clarify. I am well aware that ffmpeg internally uses ac3parse, ac3 etc. for eac3 as well as ac3 - but as the option eac3 exists, we should properly supply it. Reason: they might split them up in "some future" - that way we have it properly right at the start.

Ok reverrted to 768kbps.  @fritsch , hope this one is ready.

**Comment by fritsch** [comment]
Thanks - feel free to merge.

**Comment by HiassofT** [comment]
This broke AC3 transcoding on AC3/DTS-only setups, it now always transcodes to EAC3 with no way to disable it.

In kodi's audio settings only AC3, transcoding and DTS options are shown:
screenshot00003

EAC3 and HD formats are hidden but still seem to be enabled in settings:
```
rpi5:~ # grep eac3 .kodi/userdata/guisettings.xml 
    <setting id="audiooutput.eac3passthrough" default="true">true</setting>
```

Here's a debug log when trying to play the Fraunhofer HE-AAC sample from here: https://www2.iis.fraunhofer.de/AAC/multichannel.html
https://paste.libreelec.tv/exotic-lion.log

The log shows eac3 transcoding is used and the AES bits indicate kodi wants to output 192kHz (AES3=0x0e) but the sink only supports up to 48kHz and then things go south with lots of AE sync errors:
```
2026-04-02 12:31:24.225 T:940      info <general>: CAESinkALSA::Initialize - Opened device "hdmi:CARD=vc4hdmi0,DEV=0,AES0=0x06,AES1=0x82,AES2=0x00,AES3=0x0e"
2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::InitializeHW - Request: periodSize 2400, bufferSize 9600
2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::InitializeHW - Got: periodSize 2400, bufferSize 9600
2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::InitializeHW - Setting timeout to 200 ms
2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::GetChannelLayout - Input Channel Count: 2 Output Channel Count: 2
2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::GetChannelLayout - Requested Layout: RAW, RAW
2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::GetChannelLayout - Got Layout: RAW, RAW (ALSA: none)
2026-04-02 12:31:24.225 T:940     debug <general>: CActiveAESink::OpenSink - ALSA Initialized:
2026-04-02 12:31:24.225 T:940     debug <general>:   Output Device : vc4-hdmi-0 (vc4hdmi0)
2026-04-02 12:31:24.225 T:940     debug <general>:   Sample Rate   : 48000
2026-04-02 12:31:24.225 T:940     debug <general>:   Sample Format : AE_FMT_S16NE
2026-04-02 12:31:24.225 T:940     debug <general>:   Channel Count : 2
2026-04-02 12:31:24.225 T:940     debug <general>:   Channel Layout: RAW, RAW
2026-04-02 12:31:24.225 T:940     debug <general>:   Frames        : 2400
2026-04-02 12:31:24.225 T:940     debug <general>:   Frame Size    : 4
2026-04-02 12:31:24.225 T:939      info <general>: CAEEncoderFFmpeg::Initialize - EAC3 encoder ready
...
2026-04-02 12:31:25.480 T:939     debug <general>: ActiveAE::SyncStream - average error -392.230772 above threshold of 100.000000
2026-04-02 12:31:25.481 T:939     debug <general>: ActiveAE::SyncStream - average error -8.230772 below threshold of 30.000000
2026-04-02 12:31:26.383 T:939   warning <general>: ActiveAE - large audio sync error: -1063.923981
2026-04-02 12:31:26.483 T:939   warning <general>: ActiveAE - large audio sync error: -1160.877667
2026-04-02 12:31:26.483 T:939     debug <general>: ActiveAE::SyncStream - average error -751.250780 above threshold of 100.000000
2026-04-02 12:31:26.487 T:939     debug <general>: ActiveAE::SyncStream - average error -15.250780 below threshold of 30.000000
2026-04-02 12:31:27.235 T:939   warning <general>: ActiveAE - large audio sync error: -1000.872518
2026-04-02 12:31:27.386 T:939   warning <general>: ActiveAE - large audio sync error: -1095.852018
2026-04-02 12:31:27.536 T:939   warning <general>: ActiveAE - large audio sync error: -1192.922665
...
```

**Comment by reardonia** [comment]
> This broke AC3 transcoding on AC3/DTS-only setups, it now always transcodes to EAC3 with no way to disable it.
> 
> In kodi's audio settings only AC3, transcoding and DTS options are shown: screenshot00003
> 
> EAC3 and HD formats are hidden but still seem to be enabled in settings:
> 
> ```
> rpi5:~ # grep eac3 .kodi/userdata/guisettings.xml 
>     <setting id="audiooutput.eac3passthrough" default="true">true</setting>
> ```
> 
> Here's a debug log when trying to play the Fraunhofer HE-AAC sample from here: https://www2.iis.fraunhofer.de/AAC/multichannel.html https://paste.libreelec.tv/exotic-lion.log
> 
> The log shows eac3 transcoding is used and the AES bits indicate kodi wants to output 192kHz (AES3=0x0e) but the sink only supports up to 48kHz and then things go south with lots of AE sync errors:
> 
> ```
> 2026-04-02 12:31:24.225 T:940      info <general>: CAESinkALSA::Initialize - Opened device "hdmi:CARD=vc4hdmi0,DEV=0,AES0=0x06,AES1=0x82,AES2=0x00,AES3=0x0e"
> 2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::InitializeHW - Request: periodSize 2400, bufferSize 9600
> 2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::InitializeHW - Got: periodSize 2400, bufferSize 9600
> 2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::InitializeHW - Setting timeout to 200 ms
> 2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::GetChannelLayout - Input Channel Count: 2 Output Channel Count: 2
> 2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::GetChannelLayout - Requested Layout: RAW, RAW
> 2026-04-02 12:31:24.225 T:940     debug <general>: CAESinkALSA::GetChannelLayout - Got Layout: RAW, RAW (ALSA: none)
> 2026-04-02 12:31:24.225 T:940     debug <general>: CActiveAESink::OpenSink - ALSA Initialized:
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Output Device : vc4-hdmi-0 (vc4hdmi0)
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Sample Rate   : 48000
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Sample Format : AE_FMT_S16NE
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Channel Count : 2
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Channel Layout: RAW, RAW
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Frames        : 2400
> 2026-04-02 12:31:24.225 T:940     debug <general>:   Frame Size    : 4
> 2026-04-02 12:31:24.225 T:939      info <general>: CAEEncoderFFmpeg::Initialize - EAC3 encoder ready
> ...
> 2026-04-02 12:31:25.480 T:939     debug <general>: ActiveAE::SyncStream - average error -392.230772 above threshold of 100.000000
> 2026-04-02 12:31:25.481 T:939     debug <general>: ActiveAE::SyncStream - average error -8.230772 below threshold of 30.000000
> 2026-04-02 12:31:26.383 T:939   warning <general>: ActiveAE - large audio sync error: -1063.923981
> 2026-04-02 12:31:26.483 T:939   warning <general>: ActiveAE - large audio sync error: -1160.877667
> 2026-04-02 12:31:26.483 T:939     debug <general>: ActiveAE::SyncStream - average error -751.250780 above threshold of 100.000000
> 2026-04-02 12:31:26.487 T:939     debug <general>: ActiveAE::SyncStream - average error -15.250780 below threshold of 30.000000
> 2026-04-02 12:31:27.235 T:939   warning <general>: ActiveAE - large audio sync error: -1000.872518
> 2026-04-02 12:31:27.386 T:939   warning <general>: ActiveAE - large audio sync error: -1095.852018
> 2026-04-02 12:31:27.536 T:939   warning <general>: ActiveAE - large audio sync error: -1192.922665
> ...

@HiassofT 

Where is the E-AC3 setting in your GUI?  Can you show the top of that same screen with the speaker config?  I don't understand why the E-AC3 setting is not showing up... how "clean" is the setup on this system?  are you certain that the settings.xml file is current with the build you are using and untouched?

fwiw I have multiple AC3 and one AC3/DTS setups and all work properly.

**Comment by HiassofT** [comment]
@reardonia this is a completely clean kodi setup (moved .kodi away) running LibreELEC master with kodi c23cbf3705b58a47b93070836be1f856ba0fac16 on a RPi5.

In this setup I used my PC monitor connected through my HDFury VRROOM and let it modify the EDID to match that of a typical TV with SPDIF audio output:

image

The resulting audio part of the EDID looks like this (you can grab the full EDID from my log, it's encoded in hex in there):
```
  Audio Data Block:
    Linear PCM:
      Max channels: 2
      Supported sample rates (kHz): 48 44.1 32
      Supported sample sizes (bits): 16
    AC-3:
      Max channels: 6
      Supported sample rates (kHz): 48 44.1 32
      Maximum bit rate: 640 kb/s
  Speaker Allocation Data Block:
    FL/FR - Front Left/Right
```

Kodi automatically hides passthrough settings in the GUI that are not supported by the hardware, in this case formats that need more than 48kHz or more than 2ch - EAC3 needs 2ch 192kHz,  TrueHD and  DTS-HD need 8ch 192kHz

Here's the top of my audio settings screen, not much more to see here though
screenshot00006

**Comment by sundermann** [comment]
It might be enough to additionally check using `IsSettingVisible`

**Comment by reardonia** [comment]
> It might be enough to additionally check using `IsSettingVisible`

A few thoughts

1) Supporting HDFury scenarios seems way, way outside the norm.
2) A PC monitor that has DD support?  There is such a thing?
3) If a user has an HDFury, they can probably be expected to override an xml setting by hand
4) On the obverse, if you make this dependent on `IsSettingVisible` then you remove the ability to override with a custom setting.

I still don't understand why this disabled E-AC3 settings in the GUI. That's the core issue.  If we are relying on the audio EDID for determining AC3 and E-AC3 support, then why are we make the settings visible at all?

@HiassofT can you repro this with a real TV?  I just want to make sure this isn't an HDFury quirk issue.

**Comment by reardonia** [comment]
@sundermann  treat it like `stereoupmix` ? always tests capability and force to false if hw is not capable, ie it overrides the default.  I find the use of `IsSettingVisible` a bit hacky but it's already an idiom in the code.  thoughts?

```
diff --git a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
index 1b99676649..389e23ed10 100644
--- a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
+++ b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
@@ -2737,7 +2737,7 @@ void CActiveAE::LoadSettings()
     m_settings.passthrough = false;
   m_settings.ac3passthrough = settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_AC3PASSTHROUGH);
   m_settings.ac3transcode = settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_AC3TRANSCODE);
-  m_settings.eac3passthrough = settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_EAC3PASSTHROUGH);
+  m_settings.eac3passthrough = IsSettingVisible(CSettings::SETTING_AUDIOOUTPUT_EAC3PASSTHROUGH) ?
settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_EAC3PASSTHROUGH) : false;
   m_settings.truehdpassthrough = settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_TRUEHDPASSTHROUGH);
   m_settings.dtspassthrough = settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_DTSPASSTHROUGH);
   m_settings.dtshdpassthrough = settings->GetBool(CSettings::SETTING_AUDIOOUTPUT_DTSHDPASSTHROUGH);
```

**Comment by CrystalP** [comment]
Similar issue with Windows from the looks of it. Instead of transcoding to AC3, it looks like Kodi attempts to open the sink with 8 channels, which is not possible for an spdif device.

**Comment by sundermann** [comment]
@reardonia I think that is correct, I wonder though, if the other settings should have the same treating just for consistency

**Comment by HiassofT** [comment]
@reardonia my HDFury was configured as a simple HDMI audio extractor - identical to the cheap ~$10-20 devices that people use to get SPDIF or what TVs do if they support SPDIF out.

I don't have access to such a TV right now (my LG OLED in the livingroom supports TrueHD/Atmos and whatnot) but I found an EDID of one of our users with an audio extractor and that didn't even bother to adjust the EDID to signal AC3/DTS support, it just contained the HDMI basic audio setting:

```
  Audio Data Block:
    Linear PCM:
      Max channels: 2
      Supported sample rates (kHz): 48 44.1 32
      Supported sample sizes (bits): 24 20 16
  Vendor-Specific Data Block (HDMI), OUI 00-0C-03:
    Source physical address: 1.3.0.0
  Speaker Allocation Data Block:
    FL/FR - Front Left/Right
```

**Comment by fritsch** [comment]
We removed edid parsing and device connection checking multiple years ago due to massive false positives. While only limited code to show settings depending on the user configured strings are active. But when we switch from HDMI to spdif, XML is not touched but stays wrong.

That's why AE sometimes doubles the checks: e.g. checking number of channels vs spdif vs enabled codecs.

Checking for visible is a shortcut to get it okay ish ... what user cannot see, user most likely did not touch ... I think we have nothing better for now.

**Comment by reardonia** [comment]
> We removed edid parsing and device connection checking multiple years ago due to massive false positives. While only limited code to show settings depending on the user configured strings are active. But when we switch from HDMI to spdif, XML is not touched but stays wrong.
> 
> That's why AE sometimes doubles the checks: e.g. checking number of channels vs spdif vs enabled codecs.
> 
> Checking for visible is a shortcut to get it okay ish ... what user cannot see, user most likely did not touch ... I think we have nothing better for now.

@fritsch & @sundermann : The general issue of how to mask visibility of settings and how that should affect defaults is probably something best left to others. I think maybe for this situation the following is better. Let me know your preference.

```
diff --git a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
index 1b99676649..1cbf65bec2 100644
--- a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
+++ b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
@@ -1735,8 +1735,10 @@ void CActiveAE::ApplySettingsToFormat(AEAudioFormat& format,
     format.m_dataFormat = AE_FMT_RAW;
     format.m_sampleRate = 48000;
     format.m_channelLayout = AE_CH_LAYOUT_2_0;
-    format.m_streamInfo.m_type =
-        settings.eac3passthrough ? CAEStreamInfo::STREAM_TYPE_EAC3 : CAEStreamInfo::STREAM_TYPE_AC3;
+    format.m_streamInfo.m_type = CAEStreamInfo::STREAM_TYPE_EAC3;
+    if (!settings.eac3passthrough ||
+        !m_sink.SupportsFormat(settings.passthroughdevice, format))
+      format.m_streamInfo.m_type = CAEStreamInfo::STREAM_TYPE_AC3;
     format.m_streamInfo.m_channels = 2;
     format.m_streamInfo.m_sampleRate = 48000;
     if (mode)
```

**Comment by fritsch** [comment]
Yes. That's a good choice.

**Comment by CrystalP** [comment]
I think that's a better location for the fix.
I'd move the conditions downgrading to AC3 after the format setting, just before `if (mode)`, so that the format passed to SupportsFormat is 100% valid. Can't assume that SupportsFormat() doesn't need all fields, even though it doesn't matter in this particular case.
